src/compat/browser_compatibility_types.ts

```diff
@@ -70,17 +71,216 @@ interface ICompatDocument extends Document {
+  updateend: Event;
+  updatestart: Event;
+}
+
+/**
+ * Type-compatible with the `SourceBuffer` type (i.e. a `SourceBuffer` is a valid
+ * `ISourceBuffer`), the `ISourceBuffer` type allows to:
+ *
+ *   - re-define some attributes or methods in cases where we detected that some
+ *     platforms have a different implementation.
+ *
+ *   - list all `SourceBuffer` attributes, methods and events that are relied on
+ *     by the RxPlayer.
+ *
+ *   - Allow an easier re-definition of that API for tests or for platforms
+ *     which do not implement it.
+ */
+export interface ISourceBuffer extends IEventTarget<ISourceBufferEventMap> {
+  appendWindowEnd: number;
+  appendWindowStart: number;
+  buffered: TimeRanges;
+  timestampOffset: number;
+  updating: boolean;
+
+  abort(): void;
+  appendBuffer(data: BufferSource): void;
+  changeType(type: string): void;
+  remove(start: number, end: number): void;
+
+  onabort: ((evt: Event) => void) | null;
+  onerror: ((evt: Event) => void) | null;
+  onupdate: ((evt: Event) => void) | null;
+  onupdateend: ((evt: Event) => void) | null;
+  onupdatestart: ((evt: Event) => void) | null;
+}
+
+/** Events potentially dispatched by an `IMediaElement` */
+export interface IMediaElementEventMap {
+  canplay: Event;
+  canplaythrough: Event;
+  encrypted: MediaEncryptedEvent;
+  ended: Event;
+  enterpictureinpicture: Event;
+  error: Event;
+  leavepictureinpicture: Event;
+  loadeddata: Event;
+  loadedmetadata: Event;
+  pause: Event;
+  play: Event;
+  playing: Event;
+  ratechange: Event;
+  seeked: Event;
+  seeking: Event;
+  stalled: Event;
+  timeupdate: Event;
+  visibilitychange: Event;
+  volumechange: Event;
+  waiting: Event;
+}
+
+/**
+ * Type-compatible with the `HTMLMediaElement` type (i.e. a `HTMLMediaElement` is
+ * a valid `IMediaElement`), the `IMediaElement` type allows to:
+ *
+ *   - re-define some attributes or methods in cases where we detected that some
+ *     platforms have a different implementation.
  *
- * Note: I prefer to define my own `ICompatHTMLMediaElement` rather to extend
- * the original definition to better detect which types have been extended and
- * are not actually valid TypeScript types.
+ *   - list all `HTMLMediaElement` attributes, methods and events that are
+ *     relied on by the RxPlayer.
+ *
+ *   - Allow a re-definition of that API for tests or for platforms which do not
+ *     implement it.
  */
-interface ICompatHTMLMediaElement extends HTMLMediaElement {
+export interface IMediaElement extends IEventTarget<IMediaElementEventMap> {
+  /* From `HTMLMediaElement`: */
+  autoplay: boolean;
+  buffered: TimeRanges;
+  childNodes: NodeList | never[];
+  clientHeight: number | undefined;
+  clientWidth: number | undefined;
+  currentTime: number;
+  duration: number;
+  ended: boolean;
+  error: MediaError | null;
+  mediaKeys: null | MediaKeys;
+  muted: boolean;
+  nodeName: string;
+  paused: boolean;
+  playbackRate: number;
+  preload: string;
```

Comment:
preload can be of type:
`preload: "none" | "metadata" | "auto" | "" `

Reply:
Done